server/src/main/java/org/elasticsearch/action/admin/indices/resolve/TransportResolveClusterAction.java

```diff
@@ -0,0 +1,323 @@
+
+    private static final Logger logger = LogManager.getLogger(TransportResolveClusterAction.class);
+    private static final String TRANSPORT_VERSION_ERROR_MESSAGE = "ResolveClusterAction requires at least Transport Version";
+
+    public static final String NAME = "indices:admin/resolve/cluster";
+    public static final ActionType<ResolveClusterActionResponse> TYPE = new ActionType<>(NAME, ResolveClusterActionResponse::new);
+    private final ThreadPool threadPool;
+    private final Executor searchCoordinationExecutor;
+    private final ClusterService clusterService;
+    private final RemoteClusterService remoteClusterService;
+    private final IndexNameExpressionResolver indexNameExpressionResolver;
+    private final boolean ccsCheckCompatibility;
+
+    @Inject
+    public TransportResolveClusterAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(NAME, transportService, actionFilters, ResolveClusterActionRequest::new, EsExecutors.DIRECT_EXECUTOR_SERVICE);
+        this.threadPool = threadPool;
+        this.searchCoordinationExecutor = threadPool.executor(ThreadPool.Names.SEARCH_COORDINATION);
+        this.clusterService = clusterService;
+        this.remoteClusterService = transportService.getRemoteClusterService();
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        this.ccsCheckCompatibility = SearchService.CCS_VERSION_CHECK_SETTING.get(clusterService.getSettings());
+    }
+
+    @Override
+    protected void doExecute(Task task, ResolveClusterActionRequest request, ActionListener<ResolveClusterActionResponse> listener) {
+        // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
+        searchCoordinationExecutor.execute(ActionRunnable.wrap(listener, l -> doExecuteForked(task, request, l)));
+    }
+
+    protected void doExecuteForked(Task task, ResolveClusterActionRequest request, ActionListener<ResolveClusterActionResponse> listener) {
+        if (ccsCheckCompatibility) {
+            checkCCSVersionCompatibility(request);
+        }
+        assert task instanceof CancellableTask;
+        final CancellableTask resolveClusterTask = (CancellableTask) task;
+        ClusterState clusterState = clusterService.state();
+        Map<String, OriginalIndices> remoteClusterIndices = remoteClusterService.groupIndices(request.indicesOptions(), request.indices());
+        OriginalIndices localIndices = remoteClusterIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+
+        Map<String, ResolveClusterInfo> clusterInfoMap = new ConcurrentHashMap<>();
+        final Runnable releaseResourcesOnCancel = () -> {
+            logger.trace("clear index responses on cancellation");
+            clusterInfoMap.clear();
+        };
```

Comment:
I can likely change it to match the proposal. As mentioned in a previous comment, I patterned the structure of this transport action class on `TransportFieldCapabilitiesAction`. I think we should have (as much as possible) one defined way to define all this structural setup for TransportActions. The current system is haphazard across TransportActions, so changing it here would create yet another variant.

Reply:
Maybe then we should remove the runnable from the field caps action. I don't see how a runnable is required.

Reply:
I looked into this more closely. 

First, I don't think we want a static method, since what needs to be modified is a local variable holding state for a particular request, namely clearing this map:

```
        Map<String, ResolveClusterInfo> clusterInfoMap = new ConcurrentHashMap<>();
```

Second, what is being done here is to have a simple closure that can be called from multiple places to clear this map and do optional logging. The reason `Runnable` was chosen is that this closure takes no arguments and returns nothing. In Java, the Functional Interface for that is Runnable: https://stackoverflow.com/questions/23958814/functional-interface-that-takes-nothing-and-returns-nothing
So Runnable is useful outside of the usual context of something to run in a new/separate thread.

We could remove the Runnable and just replace `releaseResourcesOnCancel.run()` with `clusterInfoMap.clear();`. The downsides of that are:

1. You lose the optional trace level logging, which I presume someone found might be important as this is how it is done in the field-caps action, as I mentioned.
2. The name of the  closure, `releaseResourcesOnCancel`, actually gives nice context as to why we call `clusterInfoMap.clear();` in several places - context that would be more obscure without that closure.

I guess we could have a method you pass the clusterInfoMap into instead of this but that seems more clunky to me.

So my vote is to keep it. Let me know if you still want me to remove it here and in the field-caps action.

Reply:
Changed to static method in next push.